### ISSUE TITLE
TooManyOpenFiles during a stress test

### ISSUE THREAD

**Issue by sammefford** [marklogic/java-client-api#269]
Logs shared by Ajit on a separate issue:

2015-03-10 02:22:11 (Thread=2643) java.lang.RuntimeException: Failed to read file: /space/builder/linux64/b8_0/qa/testdata/4.0/non-geo-bucket/doc-1932.xml
    at com.sun.jersey.multipart.impl.MultiPartReaderClientSide.readFrom(MultiPartReaderClientSide.java:146)
    at com.sun.jersey.multipart.impl.MultiPartReaderClientSide.readFrom(MultiPartReaderClientSide.java:82)
    at com.sun.jersey.api.client.ClientResponse.getEntity(ClientResponse.java:565)
    at com.sun.jersey.api.client.ClientResponse.getEntity(ClientResponse.java:517)
    at com.marklogic.client.impl.JerseyServices.makeResults(JerseyServices.java:4448)
    at com.marklogic.client.impl.JerseyServices.getIteratedResourceImpl(JerseyServices.java:3181)
    at com.marklogic.client.impl.JerseyServices.getBulkDocumentsImpl(JerseyServices.java:832)
    at com.marklogic.client.impl.JerseyServices.getBulkDocuments(JerseyServices.java:729)
    at com.marklogic.client.impl.DocumentManagerImpl.read(DocumentManagerImpl.java:440)
    at com.marklogic.client.impl.JSONDocumentImpl.read(JSONDocumentImpl.java:26)
    at com.marklogic.client.impl.DocumentManagerImpl.read(DocumentManagerImpl.java:418)
    at com.marklogic.client.impl.JSONDocumentImpl.read(JSONDocumentImpl.java:26)
    at com.marklogic.client.impl.DocumentManagerImpl.read(DocumentManagerImpl.java:406)
    at com.marklogic.client.impl.JSONDocumentImpl.read(JSONDocumentImpl.java:26)
    at com.marklogic.client.impl.PojoRepositoryImpl.read(PojoRepositoryImpl.java:293)
    at com.marklogic.client.impl.PojoRepositoryImpl.read(PojoRepositoryImpl.java:285)
    at test.stress.JavaAPISession.countDocumentsViaPojo(JavaAPISession.java:409)
    at test.stress.PojoLoadTester.verifyLoaded(PojoLoadTester.java:60)
    at test.stress.RestLoadTester.loadContentFromDir(RestLoadTester.java:195)
    at test.stress.RestLoadTester.runTest(RestLoadTester.java:273)
    at test.stress.StressTest.run(StressTest.java:89)
Caused by: org.jvnet.mimepull.MIMEParsingException: java.io.IOException: Too many open files
    at org.jvnet.mimepull.MemoryData.createNext(MemoryData.java:98)
    at org.jvnet.mimepull.Chunk.createNext(Chunk.java:64)
    at org.jvnet.mimepull.DataHead.addBody(DataHead.java:87)
    at org.jvnet.mimepull.MIMEPart.addBody(MIMEPart.java:197)
    at org.jvnet.mimepull.MIMEMessage.makeProgress(MIMEMessage.java:240)
    at org.jvnet.mimepull.MIMEMessage.parseAll(MIMEMessage.java:181)
    at org.jvnet.mimepull.MIMEMessage.getAttachments(MIMEMessage.java:106)
    at com.sun.jersey.multipart.impl.MultiPartReaderClientSide.readMultiPart(MultiPartReaderClientSide.java:187)
    at com.sun.jersey.multipart.impl.MultiPartReaderClientSide.readFrom(MultiPartReaderClientSide.java:144)
    ... 20 more
......
......
javax.ws.rs.WebApplicationException: org.jvnet.mimepull.MIMEParsingException: java.io.IOException: Too many open files
2015-03-10 02:22:11 (Thread=2643) java.lang.RuntimeException: Failed to read file: /space/builder/linux64/b8_0/qa/testdata/4.0/non-geo-bucket/doc-1116.xml
 
2015-03-10 02:07:51 (Thread=2643) ========
Starting test with uniqueURI = /UFQFVCMIAHOTZPSS/20152102751185/4/
#
TestcaseID = pojoLoadTester
ThreadID = 2643
BinaryTest = false
JSONTest = false
InsertTime = false
NumberOfLoops = 5
MaxSleepTime = 1
LogOption = debug
WriteToScreen = true
LogFilename = auto
OutputFileName =
 
 
2015-03-10 02:07:51 (Thread=2643) About to load 2501 documents.
2015-03-10 02:12:44 (Thread=2643) About to load 2501 documents.
2015-03-10 02:12:44 (Thread=2643) About to load 2501 documents.
2015-03-10 02:21:07 (Thread=2643) About to load 2501 documents.
2015-03-10 02:22:11 (Thread=2643) java.lang.RuntimeException: Failed to read file: /space/builder/linux64/b8_0/qa/testdata/4.0/non-geo-bucket/doc-592.xml
java.lang.RuntimeException: Failed to read file: /space/builder/linux64/b8_0/qa/testdata/4.0/non-geo-bucket/doc-592.xml
    at test.stress.RestLoadTester.loadContentFromDir(RestLoadTester.java:178)
 
    at test.stress.RestLoadTester.runTest(RestLoadTester.java:273)
    at test.stress.StressTest.run(StressTest.java:89)
java.lang.RuntimeException: Failed to read file: /space/builder/linux64/b8_0/qa/testdata/4.0/non-geo-bucket/doc-2273.xml
    at test.stress.RestLoadTester.loadContentFromDir(RestLoadTester.java:178)
    at test.stress.RestLoadTester.runTest(RestLoadTester.java:273)
    at test.stress.StressTest.run(StressTest.java:89)
java.lang.RuntimeException: Failed to read file: /space/builder/linux64/b8_0/qa/testdata/4.0/non-geo-bucket/doc-2327.xml
    at test.stress.RestLoadTester.loadContentFromDir(RestLoadTester.java:178)
 
........
 
2015-03-10 02:22:12 (Thread=2643) Test took: 860953ms
Thread #4: finished loop #1 ***********************

**Comment by sammefford** [comment]
Based on a lot of investigation I'm thinking we're facing legitimate usage here and we just need to increase the allowed number of open files on linux. Right now ulimit -a returns 4096 meaning we're limited to 4096 open files for our user. Yet I'm seeing up to 3058 /tmp/MIME* files in use at once by our stress test with only 20-25 threads running. Here's my command:

echo `lsof -p 27699 | grep /tmp/MIME | wc -l` `ls -1 /tmp/MIME* | wc -l`

We sometimes run our stress test with many more threads. If I watch it, the number of /tmp/MIIME* files frequently drops back down to less than 100 indicating to me that we're appropriately closing the temp files but we sometimes need a lot at once. (I say "we" loosely, the underlying library which creates the /tmp/MIME* files is mimepull which is used by Jersey 1.17 which we're currently using). The reason we're using so many temp files is we're doing large bulk operations in parallel in multiple threads. This is a stress test, after all.

**Comment by sammefford** [comment]
I'm not sure why this got re-opened.  We changed the ulimit to 20,000 and that gave us the number of file handles we needed for this problem to go away even when running our stress tests with 100 threads.
